ykllvmwrap/src/ykllvmwrap.cc

```diff
@@ -113,19 +136,41 @@ extern "C" void *__ykllvmwrap_irtrace_compile(char *FuncNames[], size_t BBs[],
   if (TraceIdx == numeric_limits<uint64_t>::max())
     errx(EXIT_FAILURE, "trace index counter overflowed");
 
+  auto SrcMod = load_module(Context, SecPtr, SecSize);
+
+  // Get var args from start_tracing call.
+  auto Inputs = get_trace_inputs(SrcMod->getFunction(FuncNames[0]), BBs[0]);
+
+  std::vector<Type*> InputTypes;
+  for (auto Val: Inputs) {
+    InputTypes.push_back(Val->getType());
+  }
+
+  // Create function to store compiled trace.
   string TraceName = string(TRACE_FUNC_PREFIX) + to_string(TraceIdx);
   llvm::FunctionType *FType =
-      llvm::FunctionType::get(Type::getVoidTy(Context), false);
+      llvm::FunctionType::get(Type::getVoidTy(Context), InputTypes, false);
   llvm::Function *DstFunc = llvm::Function::Create(
       FType, Function::InternalLinkage, TraceName, DstMod);
   DstFunc->setCallingConv(CallingConv::C);
+
+  // Create entry block and setup builder.
   auto DstBB = BasicBlock::Create(Context, "", DstFunc);
   llvm::IRBuilder<> Builder(Context);
   Builder.SetInsertPoint(DstBB);
 
-  auto SrcMod = load_module(Context, SecPtr, SecSize);
-
   llvm::ValueToValueMapTy VMap;
+  // Map trace inputs to their appropriates variables in the trace.
+  for (size_t Idx = 0; Idx != Inputs.size(); Idx++) {
+    Value* OldVal = Inputs[Idx];
+    Value* NewVal = DstFunc->getArg(Idx);
+    if (NewVal->getType()->isPointerTy()) {
```

Comment:
I assumed this would be correct by construction? In other words we probably want:

```c++
assert(NewVal->getType()->isPointerTy());
VMap[OldVal] = NewVal;
```
```

Reply:
You're right. I had at some point considered allowing the user to pass things by value, but then realised that just muddies the waters as it allows users to write to a variable in their original code, but then the trace can't write to it, which is just confusing. So I agree we should enforce that users can only pass pointers.

Reply:
Fixed in 815ed73.